client.go

```diff
@@ -2,56 +2,154 @@ package main
-	recvChan       <-chan *arbor.ProtocolMessage
-	recieveHandler func(*arbor.ChatMessage)
 	Composer
+	address string
+	arbor.ReadWriteCloser
+	connectFunc       Connector
+	disconnectHandler func(tui.Connection)
+	receiveHandler    func(*arbor.ChatMessage)
+	stopSending       chan struct{}
+	stopReceiving     chan struct{}
+}
+
+// NewNetClient creates a NetClient configured to communicate with the server at the
+// given address and to use the provided archive to store the history.
+func NewNetClient(address, username string, history *archive.Archive) (*NetClient, error) {
+	composerOut := make(chan *arbor.ProtocolMessage)
+	stopSending := make(chan struct{})
+	stopReceiving := make(chan struct{})
+	nc := &NetClient{
+		address:       address,
+		Archive:       history,
+		connectFunc:   TCPDial,
+		Composer:      Composer{username: username, sendChan: composerOut},
+		stopSending:   stopSending,
+		stopReceiving: stopReceiving,
+	}
+	return nc, nil
+}
+
+// SetConnector changes the function used to connect to a server address. This is
+// useful both for testing purposes and to change the transport mechanism of the
+// io.ReadWriteCloser. To avoid race conditions, change this before calling
+// Connect() for the first time.
+func (nc *NetClient) SetConnector(connector Connector) {
+	nc.connectFunc = connector
+}
+
+// OnDisconnect sets the handler for disconnections. This should be done before
+// calling Connect() for the first time to avoid race conditions.
+func (nc *NetClient) OnDisconnect(handler func(tui.Connection)) {
+	nc.disconnectHandler = handler
+}
+
+// OnReceive sets the handler for when ChatMessages are received. This should be done before
+// calling Connect() for the first time to avoid race conditions.
+func (nc *NetClient) OnReceive(handler func(*arbor.ChatMessage)) {
+	nc.receiveHandler = handler
+}
+
+// Connect resolves the address of the NetClient and attempts to establish a connection.
+func (nc *NetClient) Connect() error {
+	conn, err := nc.connectFunc(nc.address)
+	if err != nil {
+		return err
+	}
+	nc.ReadWriteCloser, err = arbor.NewProtocolReadWriter(conn)
+	if err != nil {
+		return err
+	}
+	go nc.send()
+	go nc.receive()
+	return nil
+}
+
+// Disconnect stops all communication with the server and closes the connection. It invokes
+// the handler set by OnDisconnect, if there is one.
+func (nc *NetClient) Disconnect() error {
+	err := nc.ReadWriteCloser.Close()
+	if nc.disconnectHandler != nil {
+		go nc.disconnectHandler(nc)
+	}
+	go func() {
+		nc.stopSending <- struct{}{}
+		nc.stopReceiving <- struct{}{}
+	}()
+
+	return err
 }
 
-// listen monitors for messages from the server and handles them.
-func (c *Client) listen() {
-	for m := range c.recvChan {
-		switch m.Type {
-		case arbor.NewMessageType:
-			if c.recieveHandler != nil {
-				c.recieveHandler(m.ChatMessage)
-				// ask notificationEngine to display the message
-				notificationEngine(c, m.ChatMessage)
+func (nc *NetClient) send() {
```

Comment:
Are we omitting `godoc` commentary here as to not be _obivous_? That's okay if we are, just that this was a conscious decision.